ops/main.py

```diff
@@ -156,6 +156,73 @@ def _get_event_args(charm, bound_event):
     return [], {}
 
 
+class _JujuEvent:
+    """Encapsulate how to figure out what event juju wants us to run
+
+    Also knows how to run 'legacy' hooks when juju called us via dispatch.
+
+    Args:
+        charm_dir: the toplevel directory of the charm
+
+    Attributes:
+        name: the name of the event to run
+        is_dispatch: did juju call dispatch (as opposed to a hook symlink)
+    """
+    def __init__(self, charm_dir: Path):
+        self._charm_dir = charm_dir
+        self._exec_path = Path(sys.argv[0])
+        if self._exec_path.is_symlink() and self._exec_path.name != 'dispatch':
+            self._init_legacy()
+        else:
+            self._init_dispatch()
+
+    def run_any_legacy_hook(self):
+        if not self.is_dispatch:
+            # we *are* the legacy hook
+            return
+
+        dispatch_path = self._charm_dir / self._dispatch_path
+        if not dispatch_path.exists():
+            # nothing to do
```

Comment:
We might want to log this. Probably at DEBUG level.
We also probably want to check executable and not just exists()

Reply:
I added a bunch of logging (and tidied up what was there a lil')